### PR TITLE
Add armed Listener with TLS handshake timeout

### PR DESCRIPTION
**Pull request by Choraden**
This patch adds server side TLS handshake timeout. It will protect forwarders listener and additionally a handshake in mitm. 

Due to the lack of a convenient solution upstream connect is given a constant 1m timeout to complete.

### CODE REVIEW

File: http_proxy.go [hunk 1]
```diff
@@ -615,7 +618,9 @@ func (hp *HTTPProxy) listen() (net.Listener, error) {
 	case HTTPScheme:
 		return listener, nil
 	case HTTPSScheme, HTTP2Scheme:
-		return tls.NewListener(listener, hp.tlsConfig), nil
+		l := NewTLSListener(listener, hp.tlsConfig)
```
Comment (mmatczuk) [inline http_proxy.go:622]: How about we passed TLSServerConfig to NewTLSListener?

**Comment by mmatczuk** [comment]
It feels like tlshandshake could be inlined given that we have the TLSListener

**Comment by mmatczuk** [comment]
Some more context it the commit messages would be helpful

File: http_proxy.go (hunk 1 above)
Comment (Choraden) [reply http_proxy.go:622]: I decided to leave it as it is, to have logging in the same place where we load certs. 
```go
func (hp *HTTPProxy) configureHTTPS() error {
	if hp.config.CertFile == "" && hp.config.KeyFile == "" {
		hp.log.Infof("no TLS certificate provided, using self-signed certificate")
	} else {
		hp.log.Debugf("loading TLS certificate from %s and %s", hp.config.CertFile, hp.config.KeyFile)
	}

	hp.tlsConfig = httpsTLSConfigTemplate()

	return hp.config.ConfigureTLSConfig(hp.tlsConfig)
}
```

**Comment by Choraden** [comment]
> It feels like tlshandshake could be inlined given that we have the TLSListener

Listener is one thing. The other is handling handshake in mitm or SC. 
It can be later extended with proper error handling.

File: http_proxy.go (hunk 1 above)
Comment (Choraden) [reply http_proxy.go]: Eventually, I changed it.

**Comment by mmatczuk** [comment]
Now we can inline handshake in the listener.

**Comment by Choraden** [comment]
I've came to a conclusion adding handshake in the listener might not be the best option. 

The logic in the martian Proxy.Serve would cause Proxy to exit on a failed handshake. 
```go
conn, err := l.Accept()
nosigpipe.IgnoreSIGPIPE(conn)
if err != nil {
	var nerr net.Error
	if ok := errors.As(err, &nerr); ok && nerr.Temporary() {
		if delay == 0 {
			delay = 5 * time.Millisecond
		} else {
			delay *= 2
		}
		if max := time.Second; delay > max {
			delay = max
		}

		log.Debugf(context.TODO(), "temporary error on accept: %v", err)
		time.Sleep(delay)
		continue
	}

	if errors.Is(err, net.ErrClosed) {
		log.Debugf(context.TODO(), "listener closed, returning")
		return err
	}

	log.Errorf(context.TODO(), "failed to accept: %v", err)
	return err
}
```

File: net.go [hunk 2]
```diff
@@ -78,3 +81,85 @@ func Listen(network, address string) (net.Listener, error) {
 	// I asked about it here: https://groups.google.com/g/golang-nuts/c/Q1I7Viz9AJc
 	return defaultListenConfig().Listen(context.Background(), network, address)
 }
+
+type Listener struct {
```
Comment (mmatczuk) [inline net.go:85]: Let's use exported variables to configure the listener like we do with http servers for instance.

File: net.go [hunk 3]
```diff
@@ -78,3 +81,85 @@ func Listen(network, address string) (net.Listener, error) {
...
+	net.Listener
+	log              log.Logger
+	tlsCfg           *tls.Config
+	handshakeTimeout time.Duration
+
+	// OnAccept is called when a new connection is successfully accepted.
+	OnAccept func(net.Conn)
```
Comment (mmatczuk) [inline net.go:92]: An interface with the callback would be better if we are planning to have more of them.

File: http_proxy.go [hunk 4]
```diff
@@ -251,6 +251,7 @@ func (hp *HTTPProxy) configureProxy() error {
 	hp.proxy.ReadTimeout = hp.config.ReadTimeout
 	hp.proxy.ReadHeaderTimeout = hp.config.ReadHeaderTimeout
 	hp.proxy.WriteTimeout = hp.config.WriteTimeout
+	hp.proxy.MITMTLSHandshakeTimeout = hp.config.TLSServerConfig.HandshakeTimeout
```
Comment (mmatczuk) [inline http_proxy.go:254]: Move it to mitm setup

File: internal/martian/proxy.go [hunk 5]
```diff
@@ -155,6 +155,10 @@ type Proxy struct {
 	// A zero or negative value means there will be no timeout.
 	WriteTimeout time.Duration
 
+	// MITMTLSHandshakeTimeout specifies the maximum amount of time to wait for a TLS handshake for a MITMed connection.
+	// Zero means no timeout.
+	MITMTLSHandshakeTimeout time.Duration
```
Comment (mmatczuk) [inline internal/martian/proxy.go:160]: Move it adjanct to MITMFilter

**Comment by Choraden** [comment]
@mmatczuk I've addressed your comments

File: net.go [hunk 6]
```diff
@@ -78,3 +82,96 @@ func Listen(network, address string) (net.Listener, error) {
...
+}
+
+func (l *Listener) Listen() error {
+	ll, err := Listen("tcp", l.Address)
+	if err != nil {
+		return err
+	}
+
+	if rl, wl := l.ReadLimit, l.WriteLimit; rl > 0 || wl > 0 {
+		// Notice that the ReadLimit stands for the read limit *from* a proxy, and the WriteLimit
```
Comment (mmatczuk) [inline net.go:113]: Let's invert and document that.

ReadLimit is for reading form the returned connections and is shared for all connections and WriteLimit is for writing to the returned connections.

File: net.go [hunk 7]
```diff
@@ -78,3 +82,96 @@ func Listen(network, address string) (net.Listener, error) {
...
+			}
+			continue
+		}
+
+		return tc, nil
+	}
+}
+
+func (l *Listener) withTLS(c net.Conn) (net.Conn, error) {
+	tc := tls.Server(c, l.TLSConfig)
```
Comment (mmatczuk) [inline net.go:152]: use Martian/std naming conn

File: net.go [hunk 8]
```diff
@@ -78,3 +82,96 @@ func Listen(network, address string) (net.Listener, error) {
...
+
+	// OnTLSHandshakeError is called after a TLS handshake errors out.
+	OnTLSHandshakeError(*tls.Conn, error)
+}
+
+type Listener struct {
+	Address          string
+	Log              log.Logger
+	TLSConfig        *tls.Config
+	HandshakeTimeout time.Duration
```
Comment (mmatczuk) [inline net.go:98]: TLSHandshakeTimeout

File: net.go [hunk 9]
```diff
@@ -78,3 +82,96 @@ func Listen(network, address string) (net.Listener, error) {
...
+	ReadLimit        int64
+	WriteLimit       int64
+	Callbacks        ListenerCallbacks
+
+	listener net.Listener
```
Comment (mmatczuk) [inline net.go:103]: How about we embedded net.Listener?

Comment (mmatczuk) [reply net.go:103]: And remove address?

File: net_test.go [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+	"github.com/saucelabs/forwarder/log"
+)
+
+func TestListenerTLSHandshakeTimeout(t *testing.T) {
+	tlsCfg := new(tls.Config)
+	if err := (&TLSServerConfig{HandshakeTimeout: 100 * time.Millisecond}).ConfigureTLSConfig(tlsCfg); err != nil {
+		t.Fatal(err)
+	}
+
+	donec := make(chan struct{})
```
Comment (mmatczuk) [inline net_test.go:26]: use done as it's more popular variant.

File: net.go (hunk 9 above)
Comment (mmatczuk) [reply net.go:103]: nah.

File: net.go [hunk 11]
```diff
@@ -78,3 +82,96 @@ func Listen(network, address string) (net.Listener, error) {
...
+				l.Log.Errorf("Failed to close TLS connection: %v", cerr)
+			}
+			continue
+		}
+
+		return tc, nil
+	}
+}
+
+func (l *Listener) withTLS(c net.Conn) (net.Conn, error) {
```
Comment (mmatczuk) [inline net.go:151]: rename c to conn and return tls.Conn

File: net.go (hunk 6 above)
Comment (Choraden) [reply net.go:113]: It's not the scope of this PR. Here I only moved the place where we apply it. I can make an issue to invert that.

File: net_test.go [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+
+	l := Listener{
+		Address:          "[::]:0",
```
Comment (mmatczuk) [inline net_test.go:29]: Why bind to IPv6?

Tests should run on localhost as much as possible.
This way you can avoid the macOS banner to allow listening from the test process.
Please one an issue to fix that accross the board i.e. in Martian tests.

Comment (Choraden) [reply net_test.go:29]: >Why bind to IPv6?

I guess, I took it from martian tests.

File: internal/martian/proxy.go [hunk 13]
```diff
@@ -512,7 +516,15 @@ func (p *Proxy) handleMITM(ctx *Context, req *http.Request, session *Session, br
...
-		if err := tlsconn.Handshake(); err != nil {
+		var hctx context.Context
+		if p.MITMTLSHandshakeTimeout > 0 {
+			var hcancel context.CancelFunc
+			hctx, hcancel = context.WithTimeout(req.Context(), p.MITMTLSHandshakeTimeout)
+			defer hcancel()
+		} else {
+			hctx = req.Context()
+		}
+		if err := tlsconn.HandshakeContext(hctx); err != nil {
```
Comment (mmatczuk) [inline internal/martian/proxy.go:530]: Can we have var error - less ugly code.

File: internal/martian/proxy.go [hunk 14]
```diff
@@ -121,6 +121,9 @@ type Proxy struct {
 	// Implementations can return ErrConnectFallback to indicate that the CONNECT request should be handled by martian.
 	ConnectFunc ConnectFunc
 
+	// ConnectTimeout specifies the maximum amount of time to wait for a dial to complete.
```
Comment (mmatczuk) [inline internal/martian/proxy.go:124]: It's not dial it's connect with upstream.

**Comment by Choraden** [comment]
@mmatczuk I've addressed your comments and opened two issues.

File: net.go (hunk 6 above)
Comment (mmatczuk) [reply net.go:113]: ok

**Review by mmatczuk — APPROVED** [review]
LGTM

**Pull request merged**